## Problem 1: Understanding Numpy Broadcasting

a) Sum (+) with a Column Vector (2 points)

• Create a 2D array A with a shape of (4, 3).

• Create a 2D array B with a shape of (3, 1) using np.array([[0],[1],[2]]).

• Sum A and B using the + operator.

• What is the shape of the resulting array? Hint: This will produce an error. Explain why the broadcasting fails in this case.

In [33]:
import numpy as np

## a)

In [34]:
A=np.random.randn(4,3)
print(A)
B=np.array([[0],[1],[2]])
print(B)


[[-0.13750395 -1.04597421 -0.5915424 ]
 [-0.83486974  1.10159128  1.36005033]
 [-0.15803821 -1.57272468 -0.56213076]
 [ 2.15108121 -1.73612434  0.62063371]]
[[0]
 [1]
 [2]]


In [35]:
Qa=A+B

ValueError: operands could not be broadcast together with shapes (4,3) (3,1) 

This part produce an error as the following:

---------------------------------------------------------------------------

ValueError                                Traceback (most recent call last)

Cell In[6], line 1

----> 1 Qa=A+B


ValueError: operands could not be broadcast together with shapes (4,3) (3,1) 


#read this documentation: https://numpy.org/doc/stable/user/basics.broadcasting.html (Two dimensions are compatible when they are equal, or one of them is 1)

This sum will produce a error and broadcasting fails in this condition as the shape of A is (4,3) and B is (3,1). 

(4,3)

(3,1)

If we start with the right most dimensions 3 and 1 --> compatible as one of them is 1

The left dimension is 4 and 3 --> they are not equal and one of them is not 1

So one of the pairs fail--> broadcasting not possible.


-------------------------

## b)

b) Sum (+) with a Row Vector (2 points)

• Create a 2D array A with a shape of (4, 3).

• Create a 2D array C with a shape of (1, 3) using np.array([[0, 1, 2]]).

• Sum A and C using the + operator.

• What is the shape of the resulting array? How to interpret the result?


In [ ]:
A=np.random.randn(4,3)
print(A)
C=np.array([[0, 1, 2]])
print(C)

[[ 0.95853416 -1.33655695  1.09174694]
 [-1.03876889  0.7196968  -0.7127034 ]
 [-0.08548708 -0.41174499 -2.23313801]
 [ 0.5320542   0.13747036  0.33977179]]
[[0 1 2]]


In [ ]:
Qb=A+C
print(Qb)
print(np.shape(Qb))

[[ 0.95853416 -0.33655695  3.09174694]
 [-1.03876889  1.7196968   1.2872966 ]
 [-0.08548708  0.58825501 -0.23313801]
 [ 0.5320542   1.13747036  2.33977179]]
(4, 3)


Shape is --> (4,3)

The broadcasting is successful as we have a (4,3) array being added to a (1,3) array.

(4,3)

(1,3)

comparing from right, 3 and 3 are equal --> compatible

left dimensions--> one of them is 1 --> compatible (dimension 1 can be expanded to 4)

(4,3)+(1,3)=(4,3)


[[ 0.95853416 -1.33655695  1.09174694]

 [-1.03876889  0.7196968  -0.7127034 ]

 [-0.08548708 -0.41174499 -2.23313801]

 [ 0.5320542   0.13747036  0.33977179]]

      +

[0 1 2]

[0 1 2]

[0 1 2]

----------------------

## C)

c) Sum (+) with a 1D Array (2 points)

• Create a 2D array A with a shape of (3, 3).

• Create a 1D array D using np.array([0, 1, 2]).

• Sum A by D using the + operator.

• What is the shape of the resulting array? How does broadcast allows the sum between

these arrays, given that they originally have different shapes?


In [36]:
A=np.random.randn(3,3)
print(A)
D=np.array([0, 1, 2])
print(D)
print(np.shape(D))

[[ 0.70169902  2.14678016 -0.46985677]
 [ 1.46647429  0.46756827  0.45634074]
 [-0.36814959  0.67047591 -0.02084464]]
[0 1 2]
(3,)


In [37]:
Qc=A+D
print(Qc)
print(np.shape(Qc))

[[ 0.70169902  3.14678016  1.53014323]
 [ 1.46647429  1.46756827  2.45634074]
 [-0.36814959  1.67047591  1.97915536]]
(3, 3)


shape is --> (3,3)

The broadcasting is successful as we have a (3,3) array being added to a 1d array (3,).

(3,3)

(3, )    (1-dim and 3 elements)

comparing from right, 3 and 3(only 3 elements in array D) are equal --> compatible

left dimensions--> (3, ) has no other dimension  --> broadcast is done across the rows.

(3,3)+(3,)=(3,3)

[[ 0.70169902  3.14678016  1.53014323]       

 [ 1.46647429  1.46756827  2.45634074]    

 [-0.36814959  1.67047591  1.97915536]]     

      +

[0  1  2]

[0  1  2]

[0  1  2]

------